src/test/ui/liveness/liveness-upvars.rs

```diff
@@ -0,0 +1,85 @@
+// check-pass
+#![crate_type = "lib"]
+#![warn(unused)]
+
+pub fn f0() {
+    let mut last = None;
+    let mut f = move |s| {
+        last = Some(s); //~  WARN value assigned to `last` is never read
+                        //~| WARN variable `last` is assigned to, but never used
+    };
+    f("a");
+    f("b");
+    f("c");
+    dbg!(last.unwrap());
+}
+
+pub fn f1() {
+    let mut sum = 0;
+    (1..10).for_each(move |x| {
+        sum += x; //~ WARN variable `sum` is assigned to, but never used
+    });
+    dbg!(sum);
+}
+
+pub fn f2() {
+    let mut c = 0;
+    let mut d = None;
+    let mut e = None;
+
+    || {
+        move || {
+            println!("{}", c);
+            c += 1;
```

Comment:
Is the reason that we get no warning here because we assume the closure may be called again?

Reply:
Yes. The store is not dead because it might be read by the closure when called again. I added similar test cases for FnOnce and generators to show the difference.